source/pki_utils.c

```diff
@@ -1,298 +1,816 @@
+    aws_array_list_clean_up(&certificates);
+    return err;
+}
+
+void aws_release_identity(CFArrayRef identity) {
+    CFRelease(identity);
+}
+
+void aws_release_certificates(CFArrayRef certs) {
+    CFRelease(certs);
+}
+
+#endif
+
+#ifdef _WIN32
+
+#    include <aws/common/clock.h>
+
+#    include <Windows.h>
+
+#    if _MSC_VER
+#        pragma warning(disable : 4221) /* aggregate initializer using local variable addresses */
+#        pragma warning(disable : 4204) /* non-constant aggregate initializer */
+#    endif
+
+#    define CERT_HASH_STR_LEN 40
+#    define CERT_HASH_LEN 20
+
+int aws_load_cert_from_system_cert_store(const char *cert_path, HCERTSTORE *cert_store, PCCERT_CONTEXT *certs) {
+
+    char *location_of_next_segment = strchr(cert_path, '\\');
+
+    if (!location_of_next_segment) {
+        return aws_raise_error(AWS_IO_FILE_INVALID_PATH);
+    }
+
+    size_t store_name_len = location_of_next_segment - cert_path;
+    DWORD store_val = 0;
+
+    if (!strncmp(cert_path, "CurrentUser", store_name_len)) {
+        store_val = CERT_SYSTEM_STORE_CURRENT_USER;
+    } else if (!strncmp(cert_path, "LocalMachine", store_name_len)) {
+        store_val = CERT_SYSTEM_STORE_LOCAL_MACHINE;
+    } else if (!strncmp(cert_path, "CurrentService", store_name_len)) {
+        store_val = CERT_SYSTEM_STORE_CURRENT_SERVICE;
+    } else if (!strncmp(cert_path, "Services", store_name_len)) {
+        store_val = CERT_SYSTEM_STORE_SERVICES;
+    } else if (!strncmp(cert_path, "Users", store_name_len)) {
+        store_val = CERT_SYSTEM_STORE_USERS;
+    } else if (!strncmp(cert_path, "CurrentUserGroupPolicy", store_name_len)) {
+        store_val = CERT_SYSTEM_STORE_CURRENT_USER_GROUP_POLICY;
+    } else if (!strncmp(cert_path, "LocalMachineGroupPolicy", store_name_len)) {
+        store_val = CERT_SYSTEM_STORE_LOCAL_MACHINE_GROUP_POLICY;
+    } else if (!strncmp(cert_path, "LocalMachineEnterprise", store_name_len)) {
+        store_val = CERT_SYSTEM_STORE_LOCAL_MACHINE_ENTERPRISE;
+    } else {
+        return false;
+    }
+
+    location_of_next_segment += 1;
+    char *store_path_start = location_of_next_segment;
+    location_of_next_segment = strchr(location_of_next_segment, '\\');
+
+    if (!location_of_next_segment) {
+        return aws_raise_error(AWS_IO_FILE_INVALID_PATH);
+    }
+
+    /* according to the docs, this has to be the System Store by itself to open the cert store
+       successfully. There are some libraries such as libcurl that don't do this copy with just the
+       System Store. I haven't been able to make that work. I suspect mutual TLS just doesn't work
+       in curl on windows? However, this does work, so this is what we're doing.
+       As far as the 128 byte length, according to the msdn documentation here:
+       https://docs.microsoft.com/en-us/windows/desktop/SecCrypto/system-store-locations
+       this value should be more than enough. We'll add an assert just in case. */

```

Comment:
I had a really hard time parsing this. I think what you're saying is that "the store_val value has to match the computed store_path and be less than 128 bytes in length, according to the docs: $url"